File: backend/src/domain/entities/Salao.js
const { DomainError } = require('../errors/DomainError');

class Salao {
  constructor({
    id = null,
    prestadorId,
    nome,
    descricao = null,
    endereco,
    capacidadeMax,
    valorDiaria,
    areaM2 = null,
    temEstacionamento = false,
    temCozinha = false,
    permiteMusicaAoVivo = false,
    aceitaPets = false,
    regrasAdicionais = null,
    deletadoEm = null,
    criadoEm = null,
  }) {
    if (!prestadorId) throw new DomainError('prestadorId obrigatório');
    if (!nome || typeof nome !== 'string') throw new DomainError('nome obrigatório');
    if (!endereco || typeof endereco !== 'string') throw new DomainError('endereco obrigatório');
    if (!Number.isInteger(capacidadeMax) || capacidadeMax <= 0) {
      throw new DomainError('capacidadeMax deve ser inteiro positivo');
    }
    if (typeof valorDiaria !== 'number' || valorDiaria < 0 || Number.isNaN(valorDiaria)) {
      throw new DomainError('valorDiaria deve ser número não-negativo');
    }
    if (areaM2 !== null && (!Number.isInteger(areaM2) || areaM2 < 0)) {
      throw new DomainError('areaM2 deve ser inteiro não-negativo');
    }

    this.id = id;
    this.prestadorId = prestadorId;
    this.nome = nome;
    this.descricao = descricao;
    this.endereco = endereco;
    this.capacidadeMax = capacidadeMax;
    this.valorDiaria = valorDiaria;
    this.areaM2 = areaM2;
    this.temEstacionamento = !!temEstacionamento;
    this.temCozinha = !!temCozinha;
    this.permiteMusicaAoVivo = !!permiteMusicaAoVivo;
    this.aceitaPets = !!aceitaPets;
    this.regrasAdicionais = regrasAdicionais;
    this.deletadoEm = deletadoEm;
    this.criadoEm = criadoEm;
  }

  pertenceA(usuario) {
    return this.prestadorId === usuario.id;
  }

  estaAtivo() {
    return this.deletadoEm === null;
  }

  arquivar(agora = new Date()) {
    if (this.deletadoEm) {
      throw new DomainError('Salão já foi arquivado');
    }
    this.deletadoEm = agora;
  }
}

module.exports = { Salao };

File: apps/cliente/lib/domain/entities/salao.dart
import '../../core/json_utils.dart';

class Salao {
  const Salao({
    required this.id,
    required this.prestadorId,
    required this.nome,
    required this.endereco,
    required this.capacidadeMax,
    required this.valorDiaria,
    this.descricao,
    this.areaM2,
    this.temEstacionamento = false,
    this.temCozinha = false,
    this.permiteMusicaAoVivo = false,
    this.aceitaPets = false,
    this.regrasAdicionais,
  });

  final int id;
  final int prestadorId;
  final String nome;
  final String? descricao;
  final String endereco;
  final int capacidadeMax;
  final num valorDiaria;
  final num? areaM2;
  final bool temEstacionamento;
  final bool temCozinha;
  final bool permiteMusicaAoVivo;
  final bool aceitaPets;
  final String? regrasAdicionais;

  factory Salao.fromJson(Map<String, dynamic> json) => Salao(
        id: json['id'] as int,
        prestadorId: json['prestadorId'] as int? ?? 0,
        nome: json['nome'] as String? ?? 'Salão',
        descricao: json['descricao'] as String?,
        endereco: json['endereco'] as String? ?? '',
        capacidadeMax: (json['capacidadeMax'] as num?)?.toInt() ?? 0,
        valorDiaria: (json['valorDiaria'] as num?) ?? 0,
        areaM2: json['areaM2'] as num?,
        temEstacionamento: lerBool(json['temEstacionamento']),
        temCozinha: lerBool(json['temCozinha']),
        permiteMusicaAoVivo: lerBool(json['permiteMusicaAoVivo']),
        aceitaPets: lerBool(json['aceitaPets']),
        regrasAdicionais: json['regrasAdicionais'] as String?,
      );
}
